File: problem11/scala/src/test/scala/MatrixProductTest.scala
import org.scalatest.funsuite.AnyFunSuite

class MatrixProductTest extends AnyFunSuite {
  test("greatestProduct finds correct product in sample matrix") {
    val matrix = Array(
      Array(1, 2, 3, 4),
      Array(5, 6, 7, 8),
      Array(9, 10, 11, 12),
      Array(13, 14, 15, 16)
    )
    val (prod, coords) = MatrixProduct.greatestProduct(matrix, 4)
    assert(prod == 43680)
    assert(coords.contains((3, 0))) // leftmost column, down
  }
}

File: problem11/kotlin/src/test/kotlin/MatrixProductTest.kt
package org.euler.problem11

import kotlin.test.Test
import kotlin.test.assertEquals

class MatrixProductTest {
    @Test
    fun testGreatestProduct4x4() {
        val matrix = listOf(
            listOf(1, 2, 3, 4),
            listOf(5, 6, 7, 8),
            listOf(9, 10, 11, 12),
            listOf(13, 14, 15, 16)
        )
        val (result, coords) = greatestProduct(matrix, 4)
        val expected = 13 * 14 * 15 * 16
        val expectedCoords = listOf(Pair(3,0), Pair(3,1), Pair(3,2), Pair(3,3))
        assertEquals(expected, result)
        assertEquals(expectedCoords, coords)
    }
}
